### PR TITLE
[#770] Add placeholder for repositories with no author commits

### LINKED ISSUE #770: Two issues when no pre-set authors in the config files and no commits were made during the specified time period
Two issues occur when a repo with no pre-specified authors in any config files and no commits made during a pre-specified time period. When running with the following arguments `-r https://github.com/reposense/testrepo-Alpha.git https://github.com/reposense/testrepo-Delta.git -s 1/2/2019 -u 25/6/2019 -v` , the following issues occur : -

### Bug 1
Since `testrepo-Alpha` has no pre-specified authors, and no commits were made during that time period, such repos will not be shown in the report page. Also, an exception will be thrown shown below: -
image

Users will not be able to determine if anyone has made any commits into that repo during that time period. Hence, we should still show the repo (even when no authors are specified) but display a blanket message stating `no commits were made during that time period`.

### Bug 2 (related)
Colors will not be assigned to any of the file formats. This could be due to the exceptions thrown form the bug above.

### PR DESCRIPTION
**Pull request by jamessspanggg**
Fixes #770

```
A repository with no authors specified within the standalone config 
file as well as no commits made during the specified period will
not be displayed in the report page.

Users will not get the error that there were no commits made during
the specified period for this particular repository.

Let's add a placeholder with an error message of 'NO COMMITS FOUND 
WITHIN THIS PERIOD OF TIME' for repositories with no author commits.
```

### CODE REVIEW

**Comment by jamessspanggg** [comment]
The placeholder:
Screen Shot 2019-06-28 at 5 10 52 PM

**Comment by yong24s** [comment]
Heads up, #743 is also working on revamping error messages.

File: src/main/java/reposense/report/ReportGenerator.java [hunk 1]
```diff
@@ -154,17 +157,25 @@ private static void analyzeRepos(String outputPath, List<RepoConfiguration> conf
...
                 logger.log(Level.SEVERE, String.format(MESSAGE_BRANCH_DOES_NOT_EXIST,
                         config.getBranch(), config.getLocation()), e);
-                generateEmptyRepoReport(repoReportDirectory.toString());
+                generateEmptyRepoReport(repoReportDirectory.toString(), Author.NAME_FAILED_TO_CLONE_OR_CHECKOUT);
                 continue;
             }
-            analyzeRepo(config, repoReportDirectory.toString());
+
+            try {
+                analyzeRepo(config, repoReportDirectory.toString());
```
Comment (eugenepeh) [inline src/main/java/reposense/report/ReportGenerator.java]: To simplify some code, the analyzeRepo can just be combined in the try catch above.

**Comment by jylee-git** [comment]
> Heads up, #743 is also working on revamping error messages.

Should we treat repos that have no author commits made during the specified time period be considered as error? 

IMO, I don't think we should treat repos that have no commits made during the specified time period be considered as error, but rather show a placeholder instead (or we can put the text `NO COMMITS FOUND WITHIN THIS PERIOD OF TIME` to be a different, conspicuous color).

File: src/main/java/reposense/report/ReportGenerator.java [hunk 2]
```diff
@@ -151,20 +154,24 @@ private static void analyzeRepos(String outputPath, List<RepoConfiguration> conf
 
             try {
                 GitCheckout.checkout(config.getRepoRoot(), config.getBranch());
+                analyzeRepo(config, repoReportDirectory.toString());
+            } catch (NoAuthorsFoundWithCommitsException e) {
```
Comment (fzdy1914) [inline src/main/java/reposense/report/ReportGenerator.java]: `no authors found with commits for` is a little weird. I think `no authors with commits found for %s (%s) ` may be better.

File: src/main/java/reposense/report/ReportGenerator.java [hunk 3]
```diff
@@ -210,10 +217,15 @@ public static void updateRepoConfig(RepoConfiguration config) {
...
      * Find and update {@code config} with all the author identities if author list is empty.
      */
-    private static void updateAuthorList(RepoConfiguration config) {
+    private static void updateAuthorList(RepoConfiguration config) throws NoAuthorsFoundWithCommitsException {
         if (config.getAuthorList().isEmpty()) {
             logger.info(String.format(MESSAGE_NO_AUTHORS_SPECIFIED, config.getLocation(), config.getBranch()));
             List<Author> authorList = GitShortlog.getAuthors(config);
+
+            if (authorList.isEmpty()) {
+                throw new NoAuthorsFoundWithCommitsException("Unable to find author with commits in this repository.");
```
Comment (fzdy1914) [inline src/main/java/reposense/report/ReportGenerator.java]: The message `Unable to find author ..` is unnecessary if the name of the exception already implies the error and you did not use the message.

File: src/main/java/reposense/report/CommitReportJson.java [hunk 4]
```diff
@@ -24,7 +24,7 @@
     /**
      * Empty constructor to construct an empty commit report.
      */
-    public CommitReportJson() {
+    public CommitReportJson(String authorName) {
```
Comment (fzdy1914) [inline src/main/java/reposense/report/CommitReportJson.java]: Instead of `authorName`,  I prefer `displayName`. Can ask seniors' opinion.

Comment (eugenepeh) [reply src/main/java/reposense/report/CommitReportJson.java]: Yes, display name is more logically correct.

File: src/main/java/reposense/report/ReportGenerator.java [hunk 5]
```diff
@@ -53,6 +54,8 @@
     private static final String MESSAGE_MALFORMED_STANDALONE_CONFIG = "%s/%s/%s is malformed for %s (%s).";
     private static final String MESSAGE_NO_AUTHORS_SPECIFIED =
             "%s (%s) has no authors specified, using all authors by default.";
+    private static final String MESSAGE_NO_AUTHORS_FOUND_WITH_COMMITS =
```
Comment (fzdy1914) [inline src/main/java/reposense/report/ReportGenerator.java]: `MESSAGE_NO_AUTHORS_FOUND_WITH_COMMITS` -> `MESSAGE_NO_AUTHORS_WITH_COMMITS_FOUND`

**Comment by yong24s** [comment]
@jylee-git 

> IMO, I don't think we should treat repos that have no commits made during the specified time period be considered as error, but rather show a placeholder instead (or if need be, we can put the text `NO COMMITS FOUND WITHIN THIS PERIOD OF TIME` to be a different, conspicuous color).

Yes, it shouldn't be an error.

I mentioned another PR as a heads up to upcoming merge conflicts ☠️.

**Comment by jamessspanggg** [comment]
> As mentioned by @jylee-git and #598, we should change the UI of the placeholder.
> Discuss with Prof. You can work on it on another PR.

Alright noted with thanks.

**Comment by eugenepeh** [comment]
>A repository with no authors specified within the standalone config 
file as well as no commits made during the specified period will
not be displayed in the report page.

Repositories with no author specified in the configuration files as well as no commits made during the specified period do not appear in the report. 

>Users will not get the error that there were no commits made during
the specified period for this particular repository.

This is caused by a TypeError in JavaScript. When an author configuration can't be found for a repository, RepoSense will perform the identification of the authors using the commits. However, in the event that even commits cannot be found, author information will not be available and result in the error stated above.

Let's add a placeholder with an error message of 'NO COMMITS FOUND 
WITHIN THIS PERIOD OF TIME' for repositories with no author commits.

**Pull request merged**